pkg/daemon/daemon.go

```diff
@@ -119,19 +127,158 @@ func (d *daemon) Run() {
+func (d *daemon) ProceedNetworkGUID(pw *PodWrapper, networkID string) error {
+	var guidAddr guid.GUID
+	allocatedGUID, err := utils.GetPodNetworkGUID(pw.ibNetwork)
+	podNetworkID := string(pw.pod.UID) + networkID
+	if err == nil {
+		// User allocated guid manually or Pod's network was rescheduled
+		err = d.mapGUIDToPodNetwork(allocatedGUID, podNetworkID, pw.pod.UID)
+		if err != nil {
+			return err
+		}
+
+		guidAddr, err = guid.ParseGUID(allocatedGUID)
+		if err != nil {
+			return fmt.Errorf("failed to parse user allocated guid %s with error: %v", allocatedGUID, err)
+		}
+	} else {
+		guidAddr, err = d.guidPool.GenerateGUID()
+		if err != nil {
+			return fmt.Errorf("failed to generate GUID for pod ID %s, wit error: %v", pw.pod.UID, err)
+		}
+
+		allocatedGUID = guidAddr.String()
+		err = d.mapGUIDToPodNetwork(allocatedGUID, podNetworkID, pw.pod.UID)
+		if err != nil {
+			return err
+		}
+
+		if err = utils.SetPodNetworkGUID(pw.ibNetwork, allocatedGUID); err != nil {
+			return fmt.Errorf("failed to set pod network guid with error: %v ", err)
+		}
+
+		// Update Pod's network annotation here, so if network will be rescheduled we wouldn't allocate it again
+		netAnnotations, err := json.Marshal(pw.networks)
+		if err != nil {
+			return fmt.Errorf("failed to dump networks %+v of pod into json with error: %v", pw.networks, err)
+		}
+
+		pw.pod.Annotations[v1.NetworkAttachmentAnnot] = string(netAnnotations)
+	}
+
+	// used GUID as net.HardwareAddress to use it in sm plugin which receive []net.HardwareAddress as parameter
+	pw.addr = guidAddr.HardWareAddress()
+	return nil
+}
+
+// Update and set Pod's network annotation
+func (d *daemon) UpdatePodNetworkAnnotation(pw *PodWrapper, removedList *[]net.HardwareAddr) error {
+	(*pw.ibNetwork.CNIArgs)[utils.InfiniBandAnnotation] = utils.ConfiguredInfiniBandPod
+
+	netAnnotations, err := json.Marshal(pw.networks)
+	if err != nil {
+		return fmt.Errorf("failed to dump networks %+v of pod into json with error: %v", pw.networks, err)
+	}
+
+	pw.pod.Annotations[v1.NetworkAttachmentAnnot] = string(netAnnotations)
+
+	if err := d.kubeClient.SetAnnotationsOnPod(pw.pod, pw.pod.Annotations); err != nil {
+		if !strings.Contains(strings.ToLower(err.Error()), "not found") {
+			return fmt.Errorf("failed to update pod annotations with err: %v", err)
+		}
+
+		if err = d.guidPool.ReleaseGUID(pw.addr.String()); err != nil {
+			log.Warn().Msgf("failed to release guid \"%s\" from removed pod \"%s\" in namespace "+
+				"\"%s\" with error: %v", pw.addr.String(), pw.pod.Name, pw.pod.Namespace, err)
+		} else {
+			delete(d.guidPodNetworkMap, pw.addr.String())
+		}
+
+		*removedList = append(*removedList, pw.addr)
+	}
+
+	return nil
+}
+
 func (d *daemon) AddPeriodicUpdate() {
 	log.Info().Msgf("running periodic add update")
 	addMap, _ := d.watcher.GetHandler().GetResults()
 	addMap.Lock()
 	defer addMap.Unlock()
+	// Contains ALL pods' networks
 	podNetworksMap := map[types.UID][]*v1.NetworkSelectionElement{}
```

Comment:
in the refactored code, i dont see where this is being used ?
can it be removed ? 

i mean, at the begining of the call its empty and is only populated by `getIbSriovPodNetwork`
then according to the code, its being called one per pod, so you never really read from it, am i worng ?

Reply:
Actually no. It is readed at L#163 and at first time, since there is no mapping it'll be written at L#171. But map persistent over all main flow, so when next networkID and same pod will be processed, L#163 will return mapped networks.

Reply:
i see , so first call for each pod will get ->**all**<- networks for the pod.

we should do one pass to get all needed information and only then start processing the pods IMO.

Reply:
Does it really significant to have separate loop to collect all pods' networks and then other loop to iterate over same networks? Assuming network map handler, I posted in upper comment, logic gonna look smthg like:
```
loop:
    getPodInfo
    processPodInfo
```

Reply:
OK lets see how it looks once PR is updated